sling/nlp/ner/annotators.h

```diff
@@ -0,0 +1,396 @@
+
+#include "sling/base/types.h"
+#include "sling/frame/store.h"
+#include "sling/nlp/document/document.h"
+#include "sling/nlp/kb/facts.h"
+#include "sling/nlp/kb/phrase-table.h"
+#include "sling/nlp/ner/chart.h"
+#include "sling/nlp/ner/idf.h"
+#include "sling/nlp/ner/resolver.h"
+
+namespace sling {
+namespace nlp {
+
+// Span categorization flags.
+enum SpanFlags {
+  SPAN_EMPHASIS          = (1 << 0),
+  SPAN_NUMBER            = (1 << 1),
+  SPAN_NATURAL_NUMBER    = (1 << 2),
+  SPAN_UNIT              = (1 << 3),
+  SPAN_CURRENCY          = (1 << 4),
+  SPAN_MEASURE           = (1 << 5),
+  SPAN_GEO               = (1 << 6),
+  SPAN_YEAR              = (1 << 7),
+  SPAN_YEAR_BC           = (1 << 8),
+  SPAN_MONTH             = (1 << 9),
+  SPAN_WEEKDAY           = (1 << 10),
+  SPAN_CALENDAR_MONTH    = (1 << 11),
+  SPAN_CALENDAR_DAY      = (1 << 12),
+  SPAN_DAY_OF_YEAR       = (1 << 13),
+  SPAN_DECADE            = (1 << 14),
+  SPAN_CENTURY           = (1 << 15),
+  SPAN_DATE              = (1 << 16),
+  SPAN_FAMILY_NAME       = (1 << 17),
+  SPAN_GIVEN_NAME        = (1 << 18),
+  SPAN_INITIALS          = (1 << 19),
+  SPAN_DASH              = (1 << 20),
+  SPAN_SUFFIX            = (1 << 21),
+  SPAN_ART               = (1 << 22),
+};
+
+// Span markers.
+extern Handle kItalicMarker;
+extern Handle kBoldMarker;
+extern Handle kPersonMarker;
+extern Handle kRedlinkMarker;
+
+// Populate chart with phrase matches. It looks up all spans (up to the maximum
+// span length) in the alias table and adds the matches to the chart. Spans
+// cannot start or end on a stop word.
+class SpanPopulator {
+ public:
+  void Annotate(const PhraseTable &aliases, SpanChart *chart);
+
+  // Add stop word.
+  void AddStopWord(Text word);
+
+ private:
+  // Check if token is a stop word.
+  bool Discard(const Token &token) const;
+
+  // Fingerprints for stop words.
+  std::unordered_set<uint64> stop_words_;
+};
+
+// Import existing spans in the underlying document into the span chart. Dates,
+// measures, and geo coordinates are imported directly into the chart. Link
+// spans are only imported if the span text is an alias for the linked entity
+// in the alias table.
+class SpanImporter {
+ public:
+  // Initialize span importer.
+  void Init(Store *store);
+
+  // Import spans from document.
+  void Annotate(const PhraseTable &aliases, SpanChart *chart);
+
+ private:
+  // Symbols.
+  Names names_;
+  Name n_time_{names_, "/w/time"};
+  Name n_quantity_{names_, "/w/quantity"};
+  Name n_geo_{names_, "/w/geo"};
+};
+
+// Prune common words from chart. Common words are classified as words with
+// IDF (Inverse Document Frequency) below the theshold. Only lowercase words
+// are pruned from the span chart.
+class CommonWordPruner {
+ public:
+  // Prune common words using IDF dictionary.
+  void Annotate(const IDFTable &dictionary, SpanChart *chart);
+
+ private:
+  // IDF threshold for pruning single token spans.
+  static constexpr float idf_threshold = 3.5;
```

Comment:
Hmm.  This is a borderline case. I have opted for using lowercase because it is a static member, rather than the kXxx for constants. I have also always found the kXxxx convention for constant a little inconsistent with the rest of the style guide which does not permit camel case. What about using all uppercase for constants? This would make them similar to the enums which are also constants?